### PR TITLE
Home work 6

### PR DESCRIPTION
**Pull request by tosh17**
homework mvvm

### CODE REVIEW

File: app/build.gradle [hunk 1]
```diff
@@ -47,6 +47,11 @@ dependencies {
     implementation 'com.google.android.material:material:1.2.1'
     implementation 'androidx.constraintlayout:constraintlayout:2.0.4'
 
+    // ViewModel
```
Comment (rHlushan) [inline app/build.gradle]: what sense of comments?

File: app/src/main/java/ru/thstdio/aa2020/ui/detail/MoviesDetailsViewModel.kt [hunk 2]
```diff
@@ -0,0 +1,34 @@
...
+import androidx.lifecycle.MutableLiveData
+import androidx.lifecycle.ViewModel
+import androidx.lifecycle.viewModelScope
+import kotlinx.coroutines.CoroutineExceptionHandler
+import kotlinx.coroutines.launch
+import ru.thstdio.aa2020.api.Repository
+import ru.thstdio.aa2020.data.Movie
+import ru.thstdio.aa2020.ui.Navigation
+
+class MoviesDetailsViewModel(private val repository: Repository, private val router: Navigation) :
```
Comment (rHlushan) [inline app/src/main/java/ru/thstdio/aa2020/ui/detail/MoviesDetailsViewModel.kt]: thinks about router: Navigation lifecycle...
try to rotate screen...

File: app/src/main/java/ru/thstdio/aa2020/ui/detail/MoviesDetailsViewModel.kt [hunk 3]
```diff
@@ -0,0 +1,34 @@
...
+    fun loadCinemaDetail(idCinema: Long) {
+        viewModelScope.launch(exceptionHandler) {
+            try {
+                _mutableMovieState.value = repository.downloadMovie(idCinema)
+            } catch (e: Exception) {
+            }
+        }
+    }
+
+    fun onRouteBack() {
```
Comment (rHlushan) [inline app/src/main/java/ru/thstdio/aa2020/ui/detail/MoviesDetailsViewModel.kt]: what sense of moving this logic to presenter?

File: app/src/main/java/ru/thstdio/aa2020/ui/detail/MoviesDetailsViewModelFactory.kt [hunk 4]
```diff
@@ -0,0 +1,18 @@
+package ru.thstdio.aa2020.ui.detail
+
+import android.content.Context
+import androidx.lifecycle.ViewModel
+import androidx.lifecycle.ViewModelProvider
+import ru.thstdio.aa2020.api.Repository
+import ru.thstdio.aa2020.ui.Navigation
+
+
+class MoviesDetailsViewModelFactory(private val context: Context, private val router: Navigation) :
```
Comment (rHlushan) [inline app/src/main/java/ru/thstdio/aa2020/ui/detail/MoviesDetailsViewModelFactory.kt]: appContext

Comment (rHlushan) [inline app/src/main/java/ru/thstdio/aa2020/ui/detail/MoviesDetailsViewModelFactory.kt]: same about val router: Navigatio

File: app/src/main/java/ru/thstdio/aa2020/ui/detail/MoviesDetailsViewModel.kt [hunk 5]
```diff
@@ -0,0 +1,34 @@
...
+    private val _mutableMovieState = MutableLiveData<Movie>()
+    val movieState: LiveData<Movie> get() = _mutableMovieState
+
+    private val exceptionHandler = CoroutineExceptionHandler { coroutineContext, exception ->
+        println("CoroutineExceptionHandler got $exception in $coroutineContext")
+    }
+
+    fun loadCinemaDetail(idCinema: Long) {
+        viewModelScope.launch(exceptionHandler) {
+            try {
```
Comment (rHlushan) [inline app/src/main/java/ru/thstdio/aa2020/ui/detail/MoviesDetailsViewModel.kt]: what sense of try/catch?

File: app/src/main/java/ru/thstdio/aa2020/ui/list/FragmentMoviesList.kt [hunk 6]
```diff
@@ -18,36 +18,24 @@ class FragmentMoviesList : FragmentNavigation(R.layout.fragment_movies_list) {
     }
 
     private val binding: FragmentMoviesListBinding by viewBinding()
-    private val exceptionHandler = CoroutineExceptionHandler { coroutineContext, exception ->
-        Log.e("Error", "CoroutineExceptionHandler got $exception in $coroutineContext")
+    private val appContext: Context
```
Comment (rHlushan) [inline app/src/main/java/ru/thstdio/aa2020/ui/list/FragmentMoviesList.kt]: private val appContext: Context
	         get() = requireActivity().applicationContext
	         
	         to global Fragment extension?

File: app/src/main/java/ru/thstdio/aa2020/ui/list/FragmentMoviesList.kt [hunk 7]
```diff
@@ -18,36 +18,24 @@ class FragmentMoviesList : FragmentNavigation(R.layout.fragment_movies_list) {
...
-        }
+            CinemaListAdapter() { cinema -> viewModel.openDetail(cinema.id) }
+        viewModel.loadMoviesList()
+        viewModel.moviesState.observe(this.viewLifecycleOwner, this::loadMoviesList)
     }
 
-    override fun onDestroyView() {
-        scope.cancel()
-        super.onDestroyView()
+    private fun loadMoviesList(listMovies: List<Movie>) {
```
Comment (rHlushan) [inline app/src/main/java/ru/thstdio/aa2020/ui/list/FragmentMoviesList.kt]: load is correct name?

File: app/src/main/java/ru/thstdio/aa2020/ui/detail/FragmentMoviesDetails.kt [hunk 8]
```diff
@@ -88,9 +77,13 @@ class FragmentMoviesDetails : FragmentNavigation(R.layout.fragment_movies_detail
         }
     }
 
-    override fun onDestroyView() {
-        scope.cancel()
-        super.onDestroyView()
-
+    private fun showActorList(actors: List<Actor>) {
+        if (actors.isNotEmpty()) {
+            val adapter = binding.recyclerView.adapter as ActorAdapter?
```
Comment (rHlushan) [inline app/src/main/java/ru/thstdio/aa2020/ui/detail/FragmentMoviesDetails.kt]: use not
as ActorAdapter?
but
as? ActorAdapter

File: app/src/main/java/ru/thstdio/aa2020/ui/list/FragmentMoviesList.kt [hunk 9]
```diff
@@ -18,36 +18,24 @@ class FragmentMoviesList : FragmentNavigation(R.layout.fragment_movies_list) {
...
+        val adapter = binding.recyclerView.adapter as CinemaListAdapter?
```
Comment (rHlushan) [inline app/src/main/java/ru/thstdio/aa2020/ui/list/FragmentMoviesList.kt]: same about casting

File: app/src/main/java/ru/thstdio/aa2020/ui/list/MoviesListViewModel.kt [hunk 10]
```diff
@@ -0,0 +1,31 @@
...
+import androidx.lifecycle.MutableLiveData
+import androidx.lifecycle.ViewModel
+import androidx.lifecycle.viewModelScope
+import kotlinx.coroutines.CoroutineExceptionHandler
+import kotlinx.coroutines.launch
+import ru.thstdio.aa2020.api.Repository
+import ru.thstdio.aa2020.data.Movie
+import ru.thstdio.aa2020.ui.Navigation
+
+class MoviesListViewModel(private val repository: Repository, val router: Navigation) :
```
Comment (rHlushan) [inline app/src/main/java/ru/thstdio/aa2020/ui/list/MoviesListViewModel.kt]: same about Navigation

File: app/src/main/java/ru/thstdio/aa2020/ui/list/MoviesListViewModel.kt [hunk 11]
```diff
@@ -0,0 +1,31 @@
...
+        println("CoroutineExceptionHandler got $exception in $coroutineContext")
+    }
+
+    fun loadMoviesList() {
+        viewModelScope.launch(exceptionHandler) {
+            _mutableMoviesState.value = repository.downloadMovies()
+        }
+    }
+
+    fun openDetail(id: Long) {
```
Comment (rHlushan) [inline app/src/main/java/ru/thstdio/aa2020/ui/list/MoviesListViewModel.kt]: same about this fun

File: app/src/main/java/ru/thstdio/aa2020/ui/list/MoviesListViewModelFactory.kt [hunk 12]
```diff
@@ -0,0 +1,17 @@
+package ru.thstdio.aa2020.ui.list
+
+import android.content.Context
+import androidx.lifecycle.ViewModel
+import androidx.lifecycle.ViewModelProvider
+import ru.thstdio.aa2020.api.Repository
+import ru.thstdio.aa2020.ui.Navigation
+
+class MoviesListViewModelFactory(private val context: Context, private val router: Navigation) :
```
Comment (rHlushan) [inline app/src/main/java/ru/thstdio/aa2020/ui/list/MoviesListViewModelFactory.kt]: same

File: app/src/main/java/ru/thstdio/aa2020/ui/detail/FragmentMoviesDetails.kt (hunk 8 above)
Comment (tosh17) [reply app/src/main/java/ru/thstdio/aa2020/ui/detail/FragmentMoviesDetails.kt]: fix

File: app/build.gradle (hunk 1 above)
Comment (tosh17) [reply app/build.gradle]: fix

File: app/src/main/java/ru/thstdio/aa2020/ui/detail/MoviesDetailsViewModelFactory.kt (hunk 4 above)
Comment (tosh17) [reply app/src/main/java/ru/thstdio/aa2020/ui/detail/MoviesDetailsViewModelFactory.kt]: fix

File: app/src/main/java/ru/thstdio/aa2020/ui/detail/MoviesDetailsViewModel.kt (hunk 5 above)
Comment (tosh17) [reply app/src/main/java/ru/thstdio/aa2020/ui/detail/MoviesDetailsViewModel.kt]: fix

File: app/src/main/java/ru/thstdio/aa2020/ui/list/FragmentMoviesList.kt (hunk 6 above)
Comment (tosh17) [reply app/src/main/java/ru/thstdio/aa2020/ui/list/FragmentMoviesList.kt]: fix

File: app/src/main/java/ru/thstdio/aa2020/ui/detail/MoviesDetailsViewModel.kt [hunk 13]
```diff
@@ -0,0 +1,26 @@
...
+import androidx.lifecycle.LiveData
+import androidx.lifecycle.MutableLiveData
+import androidx.lifecycle.ViewModel
+import androidx.lifecycle.viewModelScope
+import kotlinx.coroutines.CoroutineExceptionHandler
+import kotlinx.coroutines.launch
+import ru.thstdio.aa2020.api.Repository
+import ru.thstdio.aa2020.data.Movie
+
+class MoviesDetailsViewModel(private val repository: Repository) :
```
Comment (rHlushan) [inline app/src/main/java/ru/thstdio/aa2020/ui/detail/MoviesDetailsViewModel.kt]: thinks about moving **idCinema: Long** to private val inside constructor

File: app/src/main/java/ru/thstdio/aa2020/ui/detail/MoviesDetailsViewModel.kt [hunk 14]
```diff
@@ -0,0 +1,26 @@
...
+    ViewModel() {
+    private val _mutableMovieState = MutableLiveData<Movie>()
+    val movieState: LiveData<Movie> get() = _mutableMovieState
+
+    private val exceptionHandler = CoroutineExceptionHandler { coroutineContext, exception ->
+        println("CoroutineExceptionHandler got $exception in $coroutineContext")
+    }
+
+    fun loadCinemaDetail(idCinema: Long) {
```
Comment (rHlushan) [inline app/src/main/java/ru/thstdio/aa2020/ui/detail/MoviesDetailsViewModel.kt]: thinks about loading this inside init block once

File: app/src/main/java/ru/thstdio/aa2020/ui/list/MoviesListViewModel.kt [hunk 15]
```diff
@@ -0,0 +1,26 @@
...
+class MoviesListViewModel(private val repository: Repository) :
+    ViewModel() {
+    private val _mutableMoviesState = MutableLiveData<List<Movie>>()
+    val moviesState: LiveData<List<Movie>> get() = _mutableMoviesState
+
+    private val exceptionHandler = CoroutineExceptionHandler { coroutineContext, exception ->
+        println("CoroutineExceptionHandler got $exception in $coroutineContext")
+    }
+
+    fun loadMoviesList() {
```
Comment (rHlushan) [inline app/src/main/java/ru/thstdio/aa2020/ui/list/MoviesListViewModel.kt]: thinks about loading this inside init block once

**Review by rHlushan — APPROVED** [review]

**Pull request merged**